File: LeetCode-Java/Generate Parentheses.java
public class Solution {
    public List<String> generateParenthesis(int n) {
        List<List<String>> recordList = new ArrayList<List<String>>();
        // first two array
        List<String> list0 = new ArrayList<String>();
        list0.add("");
        recordList.add(list0);
        List<String> list1 = new ArrayList<String>();
        list1.add("()");
        recordList.add(list1);
        for(int i = 2;i <= n;i ++)
        {
            List<String> newList = new ArrayList<String>();
            for(int j = 0;j < i;j ++)
            {
                List<String> tempList1 = recordList.get(j);
                List<String> tempList2 = recordList.get(i - j - 1);
                for(int m = 0; m < tempList1.size(); m ++)
                {
                    for(int k = 0;k < tempList2.size();k ++)
                    {
                        String tempResult = "("+tempList1.get(m) + ")" + tempList2.get(k);
                        newList.add(tempResult);
                    }
                }
            }
            recordList.add(newList);
        }
        return recordList.get(n);
    }
}

File: Leetcode-cpp/generate_parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<vector<string>> record;
        for (int i = 0; i <= n; ++i) {
            vector<string> tempRecord;
            if (i == 0) {
                tempRecord.push_back("");
            } else {
                for (int j = 0; j < i; ++j) {
                    for (int m = 0; m < record[j].size(); ++m) {
                        for (int n = 0; n < record[i - j - 1].size(); ++n) {
                            tempRecord.push_back("(" + record[j][m] + ")" + record[i - j - 1][n]);
                        }
                    }
                }
            }
            record.push_back(tempRecord);
        }
        return record[n];
    }
};